### ISSUE TITLE
[connect-web] Improved error handling for errors thrown by fetch

### ISSUE THREAD

**Issue by tstehr** [bufbuild/connect-es#715]
Hi all,

first of all thank you for making this library. We recently switched to it and it's been a huge DX improvement 👍  

**Is your feature request related to a problem? Please describe.**

I'm using `createGrpcWebTransport` from `connect-web` to create a promise client for our gRPC web backend. Any error thrown by fetch calls is wrapped in a `ConnectError` with `Code.Unknown` and no `reason` set.

For me, these errors are almost all related to a user being offline when the call is made or going offline while a stream is open. I'd like to turn them into domain-specific errors that can be handled by application logic. This is hard to do for errors with `Code.Unknown`, since they can also be thrown for other unrelated reasons that I don't want to mask.

**Describe the solution you'd like**

I'd like the errors thrown by fetch to be differentiable from other unkown errors in some way. For me using `Code.Unavailable` would work, however I could see that being problematic since not [all errors from fetch](https://developer.mozilla.org/en-US/docs/Web/API/fetch#exceptions) are technically related to availability. 

Alternatively I could see setting the `reason` property or indicating the error source in some other way.

**Describe alternatives you've considered**

I considered string matching on the error message. But this is just the message of the underlying fetch error, which varies across browsers, making this feel very brittle

For now I'm using the workaround of passing a custom `fetch` function to `createGrpcWebTransport`. This function wraps errors from fetch and turns them into a `ConnectError` with `Code.Unavailable` as follows:

```typescript
async function fetchWithConnectErrors(...args: Parameters<typeof fetch>): ReturnType<typeof fetch> {
  // handle errors during the fetch call itself
  let response;
  try {
    response = await fetch(...args);
  } catch (e) {
    if (e instanceof TypeError) {
      throw new ConnectError('Fetch error', Code.Unavailable, undefined, undefined, e);
    }
    throw e;
  }

  if (!response.body) {
    return response;
  }

  const reader = response.body?.getReader();

  // Handle errors during the response's body stream. This is needed for stream calls where the body
  // continues after the fetch already has completed
  const wrappedStream = new ReadableStream<Uint8Array>({
    start(controller) {
      async function push() {
        try {
          while (true) {
            // Read from the stream
            const { done, value } = await reader.read();
            // Exit if we're done
            if (done) {
              controller.close();
            }
            // Else yield the chunk
            controller.enqueue(value);
          }
        } catch (e) {
          if (e instanceof TypeError) {
            controller.error(
              new ConnectError('Fetch error', Code.Unavailable, undefined, undefined, e),
            );
          }
          controller.error(e);
        } finally {
          reader.releaseLock();
        }
      }

      push().catch(reportError);
    },
  });

  const wrappedResponse = new Response(wrappedStream, {
    status: response.status,
    statusText: response.statusText,
    headers: response.headers,
  });
  return wrappedResponse;
}
```

**Comment by timostamm** [comment]
Hey Tilman, I wish we could raise Code.InvalidArgument in case of an invalid request header or baseUrl, Code.Unavailable for an unreachable host, Code.PermissionDenied for a failed CORS preflight request, and so on. 

But with the current fetch API, we would have to string match for many different browsers, and we would still not be able to differentiate properly because "Failed to fetch" is often the only information that Chrome gives us.

Because we don't really know the cause, errors are wrapped with `new ConnectError(String(reason), Code.Unknown)` ([source](https://github.com/bufbuild/connect-es/blob/ef8b7fa37d56f4bd4f8d4e2133a5834776dbeed1/packages/connect/src/connect-error.ts#L122)).  This has the advantage that the type and message of the original error is available on the first line. We can't really use Code.Unavailable here because it means that clients should back off and retry if they want to, and we wouldn't want that to happen for a failed CORS preflight.

Do you think it would be an improvement for your use case to be able to distinguish an error sent by the server from an error raised by the client itself? It's a common problem, see https://github.com/bufbuild/connect-go/issues/222 and https://github.com/bufbuild/connect-go/pull/374. 

Specifically for browser going offline, you can make use of [navigator.online](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine), or the [Network Information API](https://developer.mozilla.org/en-US/docs/Web/API/Network_Information_API#browser_compatibility) if available.

**Comment by tstehr** [comment]
Thank you for the fast and comprehensive response!

> Hey Tilman, I wish we could raise Code.InvalidArgument in case of an invalid request header or baseUrl, Code.Unavailable for an unreachable host, Code.PermissionDenied for a failed CORS preflight request, and so on.
> 
> But with the current fetch API, we would have to string match for many different browsers, and we would still not be able to differentiate properly because "Failed to fetch" is often the only information that Chrome gives us.

That would in fact be awesome. Having tried the string matching route myself I understand that it's sadly not feasible to do. 

> Because we don't really know the cause, errors are wrapped with new ConnectError(String(reason), Code.Unknown) ([source](https://github.com/bufbuild/connect-es/blob/ef8b7fa37d56f4bd4f8d4e2133a5834776dbeed1/packages/connect/src/connect-error.ts#L122)).

It would be nice to also have the original error as the `cause` via `ConnectError(String(reason), Code.Unknown, undefined, undefined, reason)`. That would allow me to inspect it and treat a `TypeError` that probably originates from `fetch` differently from other client errors.

> Do you think it would be an improvement for your use case to be able to distinguish an error sent by the server from an error raised by the client itself? It's a common problem, see https://github.com/bufbuild/connect-go/issues/222 and https://github.com/bufbuild/connect-go/pull/374.

That would be helpful for sure, but probably not enough for my use case.

To go into a little more detail: This came up due to an error during protobuf serialization, where were passing malformed data into a service (with a cast that masked the typescript error). The resulting error is also an `Unknown` error and I would also consider it a client error.

In my case it would help to be able to differentiate the *layer* in the client that was causing the error (e.g. its this a transport error or a serialization error). I'm not sure thats something that makes sense for this project as a whole since it's basically a workaround for the bad `fetch` error API design.

> Specifically for browser going offline, you can make use of [navigator.online](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine), or the [Network Information API](https://developer.mozilla.org/en-US/docs/Web/API/Network_Information_API#browser_compatibility) if available.

Thank you for the pointer, I'll look into it. I anticipate that I'll still need to handle errors from fetch, since they could also happen when the user is only but the backend is temporarily offline.

**Comment by timostamm** [comment]
> In my case it would help to be able to differentiate the layer in the client that was causing the error

I understand that this can be helpful, but it adds quite a bit of complexity to the transports we want to keep as simple and small as possible - for small bundle size, but also because `fetch` is often polyfilled.

> It would be nice to also have the original error as the `cause`

We have been considering this early on. This is reasonable, let's do it.

**Comment by tstehr** [comment]
Thank you for the speed fix 👍

**Comment by timostamm** [comment]
The improvement to ConnectError.from was released in [v0.12.0](https://github.com/bufbuild/connect-es/releases/tag/v0.11.1). The `cause` property should contain the original fetch error now.
